### ISSUE TITLE
CLIPRDR don't work with RAIL

### ISSUE THREAD

**Issue by YuriSchastny** [FreeRDP/FreeRDP#2158]
Hi!

This bug is global for all versions of FreeRDP that I used. 1.0.2 from repository and fresher version from master branch.

**System info:**

OS Local: Linux Mint 16
OS Remote: Windows Server 2008 R2 SP1 Standard
FreeRDP: compiled from master branch 15.10.2014

``` bash
$ uname -a
Linux 3.11.0-12-generic #19-Ubuntu SMP Wed Oct 9 16:20:46 UTC 2013 x86_64 x86_64 x86_64 GNU/Linux
$ xfreerdp --version
This is FreeRDP version 1.2.0-beta1 (git 1.2.0-beta1+android7-874-g4c455b)
```

**Reproduce:**

Install Mint 16, install all updates, and run following commands:

``` bash
$ cd /usr/src
$ git clone git://github.com/FreeRDP/FreeRDP.git
$ cd FreeRDP
$ aptitude install libcunit1-dev libdirectfb-dev xmlto doxygen libxtst-dev libavutil-dev libavcodec-dev build-essential git-core cmake libssl-dev libx11-dev libxext-dev libxinerama-dev libxcursor-dev libxdamage-dev libxv-dev libxkbfile-dev libasound2-dev libcups2-dev libxml2 libxml2-dev   libxrandr-dev libgstreamer0.10-dev libgstreamer-plugins-base0.10-dev libxi-dev libgstreamer-plugins-base1.0-dev
$ cmake -DCMAKE_BUILD_TYPE=Debug -DWITH_SSE2=ON .
$ make
$ checkinstall make install
$ which xfreerdp
$ xfreerdp --version
```

than connect to Win2k8 r2 sp1:

``` bash
$ xfreerdp /u:User /p:Password /v:windows.local /app:"||word" +clipboard
```

copy some text localy and try paste it to remote application.

Office applications frozen or crash. Sometimes hang. Notepad do nothing (don't paste text as buffer is empty).

Only RAIL module - all OK. Only CLIPRDR module - it's ok too. Don't work CLIPRDR with RAIL. I think this bug in RAIL or CLIPRDR module. I'm betting on the RAIL :)

![2014-10-17 20 25 11](https://cloud.githubusercontent.com/assets/9254729/4682321/b58f0658-561b-11e4-8047-b47dc1f2d75e.png)
![2014-10-17 20 27 43](https://cloud.githubusercontent.com/assets/9254729/4682240/fafc42d8-561a-11e4-987c-d1616ad4cc55.png)

P.S.: Also pay attention to the terrible black border around the window in pic.1
Maybe it https://en.wikipedia.org/wiki/Remote_Desktop_Protocol#Version_8.1
`This version also fixes some visual glitches with Microsoft Office 2013 when running as a RemoteApp.`

**Comment by wapsi** [comment]
Any progress? Some branch that could be tested?

**Comment by bmiklautz** [comment]
@wapsi you can try our nightly builds an check if this is still present on master.

**Comment by wapsi** [comment]
Just tried, does not work. Remote application doesn't crash but the clipboard sync just doesn't work at all: remote app just copy/paste inside the rdp session only. When I try to copy/paste text from/to Linux side there is no any syncronization between RDP/Linux local clipboard. Used command: /usr/local/bin/xfreerdp /u:<username> /p:<password> /v:<hostname> /app:"||notepad" +clipboard

**Comment by MrObvious** [comment]
Confirmed, it's not working in the nightly builds.

**Comment by meersjo** [comment]
Have been strugging with this, too. Just built current master (bb3882b), and can confirm that clipboard does not work with /app, but does work in regular rdp mode.

**Comment by Rondom** [comment]
This seems to be a duplicate of #1636 doesn't it?

**Comment by meersjo** [comment]
@Rondom : It seems so. I believe this is a pretty important feature, which now seems to be outstanding since 2013... For me, it's the only reason I still have a citrix client installed.

**Comment by wapsi** [comment]
@meersjo :+1:  I strongly agree.

**Comment by rdlaitila** [comment]
+1 clipboard works in traditional session but not in RAIL

Version:

~$ /opt/freerdp-nightly/bin/xfreerdp --version
This is FreeRDP version 2.0.0-dev (git n/a)

Using:

/opt/freerdp-nightly/bin/xfreerdp +fonts +clipboard /cert-ignore /u:<user> /p:<pass> /app:"||cmd" /v:127.0.0.1

**Comment by peltzi** [comment]
+1

**Version (built from Github):**

```
$ xfreerdp --version
This is FreeRDP version 2.0.0-dev (git a132545)
```

**Using:**

`$ xfreerdp /d:DOMAIN /u:USER /app:'C:\Program Files (x86)\Microsoft Office\Office16\excel.exe' /v:SERVER +clipboard +fonts`

**RemoteApp Host:** Windows Server 2012 R2 with latest patches

**Comment by Simnol** [comment]
Adding my name to this as having the issue.

**Comment by blasmat** [comment]
This issue is affecting us as well.

**Comment by danlsgiga** [comment]
+1

**Comment by ShaunMaher** [comment]
Rail as advanced quite a lot since last I tested it.  Moving windows and menus popping up seems to work well.  Still being killed by lack of clipboard though.

So, +1 I guess.

**Comment by meersjo** [comment]
Just tested with current master (c325ec5) - still does not work.

**Comment by akallabeth** [comment]
Related to #1636

**Comment by boospy** [comment]
Testet with 2.0.0-dev, did not work.

**Comment by ilammy** [comment]
@boospy well, you can try out this totally untested patch which seems to do whatever I intended to do. However, I am still unable to test it due to issues with environment.

```diff
diff --git a/client/X11/xf_client.c b/client/X11/xf_client.c
index 5b5bc51..ccc1796 100644
--- a/client/X11/xf_client.c
+++ b/client/X11/xf_client.c
@@ -585,7 +585,7 @@ BOOL xf_create_window(xfContext* xfc)
 	}
 	else
 	{
-		xfc->drawable = DefaultRootWindow(xfc->display);
+		xfc->drawable = xf_CreateDummyWindow(xfc);
 	}
 
 	ZeroMemory(&gcv, sizeof(gcv));
diff --git a/client/X11/xf_event.c b/client/X11/xf_event.c
index f072021..d1a30f4 100644
--- a/client/X11/xf_event.c
+++ b/client/X11/xf_event.c
@@ -1061,10 +1061,7 @@ BOOL xf_event_process(freerdp* instance, XEvent* event)
 			break;
 	}
 
-	if (!xfc->remote_app)
-	{
-		xf_cliprdr_handle_xevent(xfc, event);
-	}
+	xf_cliprdr_handle_xevent(xfc, event);
 
 	xf_input_handle_event(xfc, event);
 	XSync(xfc->display, FALSE);
diff --git a/client/X11/xf_rail.c b/client/X11/xf_rail.c
index 5f40057..96befbb 100644
--- a/client/X11/xf_rail.c
+++ b/client/X11/xf_rail.c
@@ -67,6 +67,27 @@ void xf_rail_enable_remoteapp_mode(xfContext* xfc)
 		xfc->drawable = DefaultRootWindow(xfc->display);
 		xf_DestroyDesktopWindow(xfc, xfc->window);
 		xfc->window = NULL;
+
+		/*
+		 * XSelection protocol does not define any global keyboard as on Windows,
+		 * instead each window has its own CLIPBOARD property for clipboard content
+		 * and there is a global flag that shows which window 'holds the clipboard'
+		 * at the moment. For regular desktop sessions the session window itself
+		 *  holds the clipboard when something is copied in the remote session, but
+		 * for seamess windows there is no suitable window: the current (active,
+		 * focused) seamless window cannot be used as it can be destroyed without
+		 * closing the session; root window of the screen cannot be used too as it
+		 * is already employed as a vessel for intersession clipboard transfers.
+		 *
+		 * Therefore we create a simple dummy window for doing this job--holding
+		 * the clipboard contents for the entire seamless session.
+		 *
+		 * See also xf_create_window() in xf_client.c, a dummy window is first
+		 * created there. xf_rail_enable_remoteapp_mode() only recreates it if
+		 * autologin had failed or we had to show the session window to the user
+		 * due to some another reason.
+		 */
+		xfc->drawable = xf_CreateDummyWindow(xfc);
 	}
 }
 
@@ -75,6 +96,7 @@ void xf_rail_disable_remoteapp_mode(xfContext* xfc)
 	if (xfc->remote_app)
 	{
 		xfc->remote_app = FALSE;
+		xf_DestroyDummyWindow(xfc, xfc->drawable);
 		xf_create_window(xfc);
 	}
 }
diff --git a/client/X11/xf_window.c b/client/X11/xf_window.c
index 03ea662..5a5e0a4 100644
--- a/client/X11/xf_window.c
+++ b/client/X11/xf_window.c
@@ -336,6 +336,18 @@ static const char* get_shm_id()
 	return shm_id;
 }
 
+Window xf_CreateDummyWindow(xfContext* xfc)
+{
+	return XCreateSimpleWindow(xfc->display, DefaultRootWindow(xfc->display),
+			0, 0, 1, 1, 0, 0, 0);
+}
+
+void xf_DestroyDummyWindow(xfContext* xfc, Window window)
+{
+	if (window)
+		XDestroyWindow(xfc->display, window);
+}
+
 xfWindow* xf_CreateDesktopWindow(xfContext* xfc, char* name, int width,
                                  int height)
 {
diff --git a/client/X11/xf_window.h b/client/X11/xf_window.h
index 0a41c79..37c92ae 100644
--- a/client/X11/xf_window.h
+++ b/client/X11/xf_window.h
@@ -150,6 +150,9 @@ xfWindow* xf_CreateDesktopWindow(xfContext* xfc, char* name, int width, int heig
 void xf_ResizeDesktopWindow(xfContext* xfc, xfWindow* window, int width, int height);
 void xf_DestroyDesktopWindow(xfContext* xfc, xfWindow* window);
 
+Window xf_CreateDummyWindow(xfContext* xfc);
+void xf_DestroyDummyWindow(xfContext* xfc, Window window);
+
 BOOL xf_GetWindowProperty(xfContext* xfc, Window window, Atom property, int length,
 		unsigned long* nitems, unsigned long* bytes, BYTE** prop);
 void xf_SendClientEvent(xfContext* xfc, Window window, Atom atom, unsigned int numArgs, ...);
```

**Comment by MrObvious** [comment]
My god, it works!

**Comment by wapsi** [comment]
Confirmed, the patch works for me as well!

**Comment by boospy** [comment]
Very thanks for the fix, works perfectly here on Kubuntu 16.04 with Version: `freerdp-nightly_2.0.0+0~20170523014829.389~1.gbp48163a_amd64.deb`

**Comment by vganyn** [comment]
Work for me too on Ubuntu 16.04 with freerdp-nightly_2.0.0+0~20170523014829.389~1.gbp48163a_amd64.deb
